arch-riscv: Enable RVV run in Minor and O3 CPU

rogerchang23424 opened the pull request:
Changes in the PR:

1. Change the vset\*vl\* instructions to jump/branch family, and implement the branchTarget.
2. Move the Vl and Vtype from decoder to PCState
3. get VL, VTYPE and VLENB value from PCState
4. Remove vtype checking in construction so that the minor and o3 cpu and decode the instructions after the vset\*vl\*

rogerchang23424 committed:
arch-riscv: Remove checked_type in StaticInst Constructor

We should not try to check vtype when decoding the instruction.
It should be checked in vset{i}vl{i} since the register can be
modified via vset{i}vl{i}

Change-Id: I403e5c4579bc5b8e6af10f93eac20c14662e4d2d

rogerchang23424 committed:
arch-riscv: Allow Minor and O3 CPU execute RVV

Change-Id: I4780b42c25d349806254b5053fb0da3b6993ca2f

rogerchang23424 commented:
Related issue: #144
To enable run RVV in minor, #225 is required

BobbyRBruce reviewed (dismissed):
Loving the progress we're making on RVV! Great work!

BobbyRBruce commented:
I'll give this another day to give others a chance to see this. Looks good to me but I'm not a RISC-V expert.

@hnpl : Can you look this over?

rogerchang23424 commented:
cc: @vmskv @adriaarmejach @alvaromorenomv

aarmejach commented:
The code LGTM to enable `minor` and `o3` now. But I have a questions and a concern:

Why do we need to replicate the `vtype` and  `vl` register state in decode? I think that using the `MISCREG` architectural values will simplify things. 

My concern is that with this path the issue #144 might still trigger. Say you execute and commit a `vsetvl` instruction, changing `vtype` and `vl` state, and then the `vsetvl` instructions is squashed due to a previous branch missprediction. The newly decoded instructions after recovering from the missprediction will see the modified `vtype` and `vl` state. Maybe I am wrong, but this might still be an issue.

I am all in on enabling Minor and O3 ASAP, so if #144 is easy to fix in this patch that would be great. 

Since we are discussing, I am interested in trying to implement a proposal that does not always squash after. Making `vsetvl`instructions behave like branches that update `vtype` and `vl` state at commit time make things quite easy. If at commit time `new_vtype != previous_vtype or new_vl != previous_vl` you squashAll (method already exists), otherwise do nothing. With this approach decode can use the current architectural state for `vtype` and `vl`. This will enable loops that have the same `vsetvl` instruction, common case, to execute much faster.

rogerchang23424 commented:
> The code LGTM to enable `minor` and `o3` now. But I have a questions and a concern:
> 
> Why do we need to replicate the `vtype` and `vl` register state in decode? I think that using the `MISCREG` architectural
> values will simplify things.

It's better to remove the `vtype` and `vl` in decoder to make it more simply. However, the `StaticInst` of vector instructions require the `vtype` and `vl` parameters to decode(construct) them.

> My concern is that with this path the issue #144 might still trigger. Say you execute and commit a `vsetvl` instruction, changing `vtype` and `vl` state, and then the `vsetvl` instructions is squashed due to a previous branch missprediction. The newly decoded instructions after recovering from the missprediction will see the modified `vtype` and `vl` state. Maybe I am wrong, but this might still be an issue.

I can't guarantee the issue #144 will be solved after the change applied, I have tested the vector with https://github.com/ksco/riscv-tests/tree/spike/isa/rv64uv, [rvv-intrinsic-doc](https://github.com/riscv-non-isa/rvv-intrinsic-doc)(all of examples are passed, except for str* since it require `vleff*` be implemented ) and [riscv_o3_bad_vsetvl_execution_reproducer](https://github.com/vmskv/gem5/tree/vmskv/rvv_bad_o3_vsetvl_execuion/riscv_o3_bad_vsetvl_execution_reproducer),all of them are passed. The example of #144 shows the correct result for both atomic and o3 cpu.

> I am all in on enabling Minor and O3 ASAP, so if #144 is easy to fix in this patch that would be great.
> 
> Since we are discussing, I am interested in trying to implement a proposal that does not always squash after. Making `vsetvl`instructions behave like branches that update `vtype` and `vl` state at commit time make things quite easy. If at commit time `new_vtype != previous_vtype or new_vl != previous_vl` you squashAll (method already exists), otherwise do nothing. With this approach decode can use the current architectural state for `vtype` and `vl`. This will enable loops that have the same `vsetvl` instruction, common case, to execute much faster.

OK, I will change the vset\*vl\* instruction to branch like instruction,  I have saw the implementation before https://github.com/hnpl/plct-gem5-1/commits/rvv-cpu.

rogerchang23424 commented:
Rebase only

hnpl commented:
>> The code LGTM to enable minor and o3 now. But I have a questions and a concern:
>>
>> Why do we need to replicate the vtype and vl register state in decode? I think that using the MISCREG architectural
values will simplify things.

> It's better to remove the vtype and vl in decoder to make it more simply. However, the StaticInst of vector instructions require the vtype and vl parameters to decode(construct) them.

I think having the decoder maintaining its own state of vtype and vl regs is okay because it looks like the RISCV decoder does not have access to the architectural state of the core, i.e. the core's ThreadContext (tc) or ExecContext (xc). 

> OK, I will change the vset*vl* instruction to branch like instruction, I have saw the implementation before https://github.com/hnpl/plct-gem5-1/commits/rvv-cpu.

I want to note that this implementation takes a different direction than having the vtype and vl states in decoder.
Instead of having vtype and vl in decoder, the vtype and vl of an instruction are stored in the instruction's PCState, which is a part of every DynInst.

Having vtype and vl in PCState is necessary if the vconfig instructions are treated as branch instructions. This is because for branch instructions, the core will determine whether the branch instruction is correctly executed by comparing the PCStates. This is reflected in this change here, https://github.com/hnpl/plct-gem5-1/commit/b061650493901143ccf60585ef87e46687e2949a#diff-366ad84d19b6100d495edd74366ab39abe982fb2265e611baa839aa9d513b3d9R101.

This is how the vtype and vl are updated,
https://github.com/hnpl/plct-gem5-1/commit/3013eb0c5738bdc588d4d830cf33351d562176a7#diff-0c78d5a93acb1d90c9296beb8cfee1f57b97aec841507f8681158cb529168fa5R90

This is where the next vector instruction acquires the vtype and vl from the previous instruction,
https://github.com/hnpl/plct-gem5-1/commit/8d946c742ea093445e406e518a7e6b485e0f2a75#diff-93a5a8581aba54897a09350ea3f735d79ef1fd441f1970010c765647f6fb4489R130

rogerchang23424 commented:
Thanks @hnpl

rogerchang23424 committed:
arch-riscv: Refactor PCState class

Change-Id: I1d25350ba2a3c7c366f42340c20b4488c33cde6f

rogerchang23424 committed:
arch-riscv: Add vlenb, vtype and vl in PCState

Change-Id: I7c2aed7dda34a1a449253671d7b86aa615c28464

rogerchang23424 committed:
arch-riscv: Change the instruction family to jump like

The method that get the vl, vtype from PCState in the next changes

Change-Id: I022b47b7a96572f6434eed30dd9f7caa79854c31

rogerchang23424 committed:
arch-riscv: Change the implementation of vset*vl*

The changes includes:

1. Add VL, Vtype and VlenbBits operands
2. Change R/W methods of VL, Vtype and VlenbBits from PCState

Change-Id: I0531ddc14344f2cca94d0e750a3b4291e0227d54

rogerchang23424 committed:
arch-riscv: Get vl, vtype and vlenb from PCState

Change-Id: I0ded57a3dc2db6fcc7121f147bcaf6d8a8873f6a

rogerchang23424 committed:
arch-riscv: Implement the branchTarget for vset*vl*

Change-Id: I10bf6be736ce2b99323ace410bff1d8e1e2a4123

rogerchang23424 committed:
arch-riscv: Remove check vconf done implementation

Change-Id: If633cef209390d0500c4c2c5741d56158ef26c00

aarmejach commented on src/arch/riscv/isa.cc:
```diff
@@ -503,9 +504,26 @@ ISA::readMiscReg(RegIndex idx)
+            if (_rvType == RV64) {
+                val &= ~(1 << 31);
+                val |= (vtype.vill << 63);
+            }
```
This made me realize that the current VTYPE BitUnion is defined as 32bit. I think RVV requires at least RV64, so VTYPE should be at least 64 bits. This would also simplify this code, as the cast to RegVal would be enough.

rogerchang23424 replied:
Change VTYPE to BitUnion64

aarmejach commented on src/arch/riscv/isa.cc:
```diff
@@ -503,9 +504,26 @@ ISA::readMiscReg(RegIndex idx)
+            VTYPE vtype = rpc.vtype();
+            val = (RegVal)vtype;
+            if (_rvType == RV64) {
+                val &= ~(1 << 31);
```
If we keep this line I'd change the `~(1 << 31)` for the actual mask or `~(1U << 31)`. Just to avoid issues on 32-bit systems or compilers interpreting it as a 64-bit constant.

rogerchang23424 replied:
Change VTYPE to BitUnion64

aarmejach commented on src/arch/riscv/pcstate.hh:
```diff
@@ -56,15 +57,28 @@ constexpr enums::RiscvType RV64 = enums::RV64;
+    typedef GenericISA::UPCState<4> Base;
+
     bool _compressed = false;
     RiscvType _rvType = RV64;
```
Should this be initialized with `isa->rvType()`?

rogerchang23424 replied:
It has initialized in isa.newPCState https://github.com/gem5/gem5/blob/stable/src/arch/riscv/isa.hh#L90

aarmejach commented on src/arch/riscv/pcstate.hh:
```diff
@@ -56,15 +57,28 @@ constexpr enums::RiscvType RV64 = enums::RV64;
     bool _compressed = false;
     RiscvType _rvType = RV64;
+    uint64_t _vlenb = 32;
+    VTYPE _vtype = 0x80000000; // vtype.vill = 1 at initial;
```
If VTYPE BitUnion is changed to 64 bits this needs to be updated

aarmejach reviewed (commented):
This is great! I was suggesting to go this way on a future iteration :)

Overall this LGTM, I have left a few minor coments.

rogerchang23424 committed:
arch-riscv: Change VTYPE to BitUnion64

Change-Id: I7620ad1ef3ee0cc045bcd02b3c9a2d83f93bf3fe

rogerchang23424 committed:
arch-riscv: Refactor the execution part of vset*vl*

Change-Id: Ie0d9671242481a85bb0fe5728748b16c3ef62592

rogerchang23424 committed:
util: Update the RISC-V PCState checkpoint

Change-Id: I64b6a3e1706173a001f5f8fb06756bd50d65f5bd

rogerchang23424 commented:
```
Logging call to command: /__w/gem5/gem5/build/ALL/gem5.opt -d /tmp/gem5outppd_thks -re --silent-redirect /__w/gem5/gem5/configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py
Starting Test Suite: test-gem5-library-riscv-hello-restore-checkpoint-ALL-x86_64-opt 
Starting Test Case: test-gem5-library-riscv-hello-restore-checkpoint-ALL-x86_64-opt
Test: test-gem5-library-riscv-hello-restore-checkpoint-ALL-x86_64-opt Failed
Starting Test Case: test-gem5-library-riscv-hello-restore-checkpoint-ALL-x86_64-opt-MatchRegex
Test: test-gem5-library-riscv-hello-restore-checkpoint-ALL-x86_64-opt-MatchRegex Failed
```

@BobbyRBruce, We need to run `util/cpt_upgrader.py` to update the `riscv-hello-example-checkpoint` resource

aarmejach reviewed (approved):
LGTM but it'd be great if we can get another +1.

powerjg reviewed (approved):
Wow! Way simpler than I expected!

Thanks for putting this together!

powerjg commented:
@BobbyRBruce, we're just waiting on you to update the checkpoints in gem5-resources. Let me know if I can help.

BobbyRBruce reviewed (approved):
Sorry for the delay. The resource is updated and the tests should pass now.
